### PR TITLE
Improved handling of large amounts of blocks

### PR DESCRIPTION
**Pull request by jacobsjo**
Rendering very large structures might not have been the primary goal of this project. Nonetheless, it does pose a problem. For example if you want to display some of the vanilla bastion elements only a some blocks are drawn. This is because limit of vertices drawn in a single webGL draw call is reached. This pull request does two things to improve this situation:

### Split StructureBuffers:
If the buffers are getting too large, they are saved and new buffers are created. During drawing separate draw calls are made for each set of buffers. The size limit is given by ```facesPerBuffer```. As this probably depends on hardware availability it is an optional parameter to the ```StructureRenderer```. The default value of 5500 probably needs some tuning.

### Culling
Similarly to how it is done in the game, faces that have a ```cullface``` set and where in the specified direction there is a solid block adjacent are not drawn. This requires efficient block lookup. Therefore I have modified the ```Structure``` class to provide constant time lookup. Also, I have not been able to find an elegant solution to determining whether a block is solid or not. For now, it is simply done by checking the name to a long hard coded list (in ```SolidHelper.js```).

For some of the helper function I have not been sure where best to place it. If you have other preferences I am happy to do some refactoring.

Culling might in the future also be useful to improve rendering of transparent objects and fluids.

### CODE REVIEW

File: render/src/StructureRenderer.ts [hunk 1]
```diff
@@ -118,7 +119,8 @@ export class StructureRenderer {
     private blockDefinitionProvider: BlockDefinitionProvider,
     private blockModelProvider: BlockModelProvider, 
     private blockAtlas: BlockAtlas,
-    private structure: StructureProvider
+    private structure: StructureProvider,
+    private facesPerBuffer: number = 5500
```
Comment (misode) [inline render/src/StructureRenderer.ts:123]: For compatibility, this should be in an options object

File: render/src/StructureRenderer.ts [hunk 2]
```diff
@@ -167,15 +169,36 @@ export class StructureRenderer {
...
+
+      positions = []
+      textureCoordinates = []
+      tintColors = []
+      blockPositions = []
+      indices = []
+      indexOffset = 0
+    }
+
+    var pushBuffers = (buffers: any, pos: vec3) => {
```
Comment (misode) [inline render/src/StructureRenderer.ts:201]: This and `finishBuffers` should be either a `function` or `const`.

File: render/src/Util.ts [hunk 3]
```diff
@@ -30,3 +31,102 @@ export function transformVectors(array: Float32Array, transformation: mat4) {
...
+        "up": "south",
+        "down": "north"
+      }
+  }
+
+  const xRotCull : Cull = {}
+  for (const d of directions) xRotCull[xMapping![d]] = yRotCull[d]
+
+  return xRotCull
+}
```
Comment (misode) [inline render/src/Util.ts:132]: Missing newline

**Review by misode — COMMENTED** [review]
Amazing! Just a few minor changes.

**Comment by jacobsjo** [comment]
I have updated the pull request according to your suggestions. One thing I just thought about is whether it might make more sense to put the solid check into the ```ResourceManager```. That would make it more compatible between Minecraft versions. Indeed there are already blocks in the 1.17 snapshots that aren't currently in the ```SolidHelper``` list

**Comment by misode** [comment]
Are you thinking something like an `OpaqueProvider`?

**Comment by jacobsjo** [comment]
Yes, a ```OpaqueProvider``` is probably the best way to do it. However, then I am not sure how best to pass it to the ```StructureRenderer``` in a compatibility way.

**Comment by misode** [comment]
What about this?
```ts
type Resources = {
  opaqueBlocks: OpaqueProvider
  blockDefinitions: BlockDefinitionProvider
  blockModels: BlockModelProvider
  blockAtlas: BlockAtlas
}

class StructureRenderer {
  constructor(
    gl: WebGLRenderingContext,
    structure: StructureProvider,
    resources: Resources,
    options?: StructureRendererOptions
  )
}
```

**Comment by jacobsjo** [comment]
All right, done.

Also, I have moved the ```solidHelper``` to the demo (as ```opaqueHelper```), so the demo uses that in the ```ResourceManager``` implementation.

**Comment by misode** [comment]
Looks good!

**Pull request merged**